Checkpoint2 Submission

lewisemm opened the pull request.

lewisemm committed:
Classes Room, Living, Office, Person, Staff and Fellow defined.

lewisemm committed:
Initialize Amity with 20 rooms. More comments. Add people functionality in Office. Add sanity checks when rooms are full.

lewisemm committed:
Apply abstraction. Create class definitions for Room, Person, Living, Office, Staff and Fellow on separate files.

lewisemm committed:
Define classes in separate files. Use import to subclass related classes.

lewisemm committed:
Create custom exception to assist in error handling.

lewisemm committed:
Ignore generated compile files (pyc).

lewisemm committed:
Refine user menu. Also add menu for staff and fellow room allocation.

lewisemm committed:
Add Amity class. Initialize Living and Office spaces here.

lewisemm committed:
Add allocating room to fellow sub-menu.

lewisemm committed:
Create constructor. Restructure code. Import necessary files.

lewisemm committed:
Allocate fellow via app option. Import missing modules. Restructure code - more abstraction. Specify ValueError at generic catch block.

lewisemm committed:
Space lines of code.

lewisemm committed:
Add constructor to initialize variables in class.

lewisemm committed:
Utilize OverflowException.

lewisemm committed:
Space lines of code.

lewisemm committed:
Initialize variables on instantiation of class object. Get class names from allocation.txt file. Assign at random. Put allocate and persist logic in place.

lewisemm committed:
Add logic retrieve object status from previous allocation operations.

lewisemm committed:
Add getter for occupants list.

lewisemm committed:
Create constructor. Add getters for max and occupants variables.

lewisemm committed:
Create constructor with name argument.

lewisemm committed:
Create file to store cross functional logic.

lewisemm committed:
Create data folder to place app's persistent files.

lewisemm committed:
Add docstrings to methods. Include mechanism to separate Fellows from Staff in file. Randomize space allocation in dynamic manner instead of constant 10 rooms.

lewisemm committed:
Abstract user menus to utilities.py file. Add allocation to office space.

lewisemm committed:
Sample allocation data.

lewisemm committed:
Add feature to print allocation list.

lewisemm committed:
Get unallocated Staff and Fellows from amity class.

lewisemm committed:
Add logic to print unallocated people.

lewisemm committed:
Reset allocations.

lewisemm committed:
Add function to get current dir.

lewisemm committed:
Add getters for livingRooms and officeRooms.

lewisemm committed:
Allocate staff through text file.

lewisemm committed:
Allocate Fellow to office via text file.

lewisemm committed:
Allocate living spaces to fellows through text file.

lewisemm committed:
Print occupants in a room. Print unallocated people.

lewisemm committed:
Add menu to select between Offices and Living spaces.

lewisemm committed:
Create package folder structure.

lewisemm committed:
Adjust syntax for class importation as module.

lewisemm committed:
Create tests.

lewisemm committed:
Apply PEP8 styling.

lewisemm committed:
Apply restructured modules when importing.

Commit:
Abstract printAllocation to a function and create it in global scope.

Commit:
Correct misleading menu.

Commit:
Relocate functions to acheive global scale.

Commit:
Fixed imports.

Commit:
Add files whose changes are to be ignore.

Commit:
Relocate method printUnallocated to local namespace.

Commit:
Add tests to check unallocated and allocated people before changes.

Commit:
Ignore all files from data folder.

lewisemm committed:
Add docstring to all functions.

lewisemm committed:
Merge branch 'develop'

Merges develop to master for checkpoint submission.

lewisemm committed:
Modify spacing to Python style.

lewisemm committed:
Merge pull request #1 from andela-lkabui/develop

Merge develop to master branch

Commit:
Add content to readme file.

Commit:
README.md updated.

lewisemm committed:
Update README.md

lewisemm committed:
Update README.md

Commit:
Rename test file for purposes of test discovery.

Commit:
Use TextTestRunner to run tests.

lewisemm committed:
Update README.md to review suggestions

- Single backticks used for inline shells.
- Change instances of the word 'folder' to 'directory'.
- Use shells for python keywords.
- Rename test file to follow 'test_*.py' convention to enable test discovery without superfluous switches.

lewisemm committed:
Apply PEP8 Style guide

- import standard lib modules ahead of thrid party and custom libs
- change function and method names to use underscore naming
- two blank lines before class definition and blank line between function/methid definitions
- getters and setters removed
- space eliminated between number and menu title when checking occupants per room

lewisemm committed:
Retrieve file names dynamically

- Absolute file paths used in all reads and writes to file
- Modify method to get current path in 'utilities.py' file

lewisemm committed:
Change directory structure

- Change directory structure to accomodate all checkpoints in one repo.

lewisemm committed:
Create models. Handle get and post bucketlists. No pagination

lewisemm committed:
Create requirements text file

- Enable use of pip install -r requirements.txt for easy project setup on different machines

lewisemm committed:
Create user registration resource

- Create user model with provosion for hashed password
- Create resource handler for user registration url
- Update requirements.txt with PassLib package

lewisemm committed:
Add authentication through tokens

- Create methods in User model to generate and verify tokens
- Grant access to endpoints based on login status

lewisemm committed:
Add authentication functionality to REST api

- Use flask-login to authenticate users and provide sessions
- Use token based authentication and eliminate use of HTTPAuth
- Create secret key as environment variable for use when creating and deciphering tokens
- Create html templates for use during login, logout and registration
- Update requirements.txt file

lewisemm committed:
Add more API endpoints

- Added routes to create bucketlist with username of user currently logged in
- Added routes to get, put and delete single bucket list
- Added routes to post single item
- Added route to put and delete single item
- Altered route to endpoints as per specification
- Require authentication to endpoints as per specification

lewisemm committed:
JSONify API's input and output

- Authentication via json input
- Output via json output
- Parse json POSTS from request body

lewisemm committed:
Dirty

lewisemm committed:
Implement authentication through tokens

- Modify verify_password to check for token availability in request header. Denies access of token is invalid/non-existent
- Annotate methods with '@auth.login_required' to deny access to users who haven't logged in
- Alter verify_password in the User's model to accept an extra argument (manager, a session) to facilitate querying of db.
- Trim superfluous packages from requirements.txt file

lewisemm committed:
Create Readme file for the checkpoint

- Includes;
    - An introduction section explaining purpose of the API
    - A section detailing routes, their corresponding HTTP methods and their behaviour

lewisemm committed:
(feature) Add more functionality to the REST API

- Eliminated the Flask-Login package to avoid the use of sessions
- Added feature to show items already existing in bucketlist when a bucketlist id is specified
- Added the feature to allow updating of items in a bucketlist
- Added the feature to allow deletion of items in a bucketlist
- Updated the requirements.txt file to reflect changes in package requirements.

lewisemm committed:
(chore) Merge branch 'checkpoint2' into checkpoint2-dev

- include file addition (readme.md)

lewisemm committed:
(feature) Add method to limit put and delete operations to bucketlist owners

Created method "is_bucketlist_owner" to determine if the current token belongs to the user who created the bucketlist before any put and delete operations are performed.

lewisemm committed:
(feature) Use Model serializations in get responses

- Make use of Flask-RESTful library to serialize models and display representations
- Update readme file
- Adjust BucketList model to include one to many unidirectional relationship for use during serialization.

lewisemm committed:
(feature) Add tests for URL

lewisemm committed:
(feature) Include more http status codes in responses

- Eliminate the use of jsonify to allow the return of status code values
- Refactor login logic to produce different informative responses for when username and/or password missing
- Eliminating the use of method views in favour of class views provided by Flask-HttpAuth library
- Refactoring logic to correct failing tests

lewisemm committed:
(feature) Finish replacing function based views with class based views from Flask-HttpAuth

- Replace all instances of 'jsonify' usage to allow returning of http status codes along with response objects
- Update requirements file with latest package requirements

lewisemm committed:
(feature) Update readme file

- Update readme file with latest package structure and added functionality

lewisemm committed:
(chore) - Change folder structure

- Changed folder structure for better code organisation
- Fixed import errors resulting from change in folder structure

lewisemm committed:
(feature) Implement search bucketlist by name feature.

- Added url to bucketlist resource to allow searching by name if the parameter is provided by the client

lewisemm committed:
(chore) Correct typos and markdown errors in readme file

lewisemm committed:
(fix) Refactor search by text feature to correctly meet specifications in checkpoint

lewisemm committed:
(feature) Implement pagination when getting a list of bucketlists

- Restructured class BucketList to allow for pagination of bucketlist items
- Added url "/bucketlists/pages/<int:pages>" to access results through pages
- Restructured classes to handle multiple return response values by using "marshal" in place of "marshal_with" (both from Flask-RESTful). Also created constructors in these classes to allow for initialization and access of the marshal fields from within other sections of the class.

lewisemm committed:
(feature) Add pagination to API

- Added pagination logic to group content into pages as specified by user. Default pagination of 20 items per page (unless user specifies otherwise)
- Added library dependency (sqlalchemy-paginator) to paginate queries
- Commented out pagination unittest because of unresolved problem. To be revisited later.

lewisemm committed:
(feature) Add .travis.yml config file

lewisemm committed:
(chore) Update requirements.txt file

lewisemm committed:
(chore) Update .gitignore file

lewisemm committed:
(chore) Add APP_SETTINGS and SECRET_KEY environment variables on travis config file

lewisemm committed:
(chore) Configure SECRET_KEY environment variable on travis config file

Added mysql as the test database on Travis CI. Also included database url in Travis CI config file.

lewisemm committed:
(chore) Add Travis CI badge to readme

- Also onfigured Travis CI with coveralls to produce test coverage badges

lewisemm committed:
(feature) Add badge for coveralls

lewisemm committed:
(feature) Create accurate test coverage reports

- Removed the virtual environment and tests folder from the test coverage report to give accurate indication of lines of code being tested.
- Created .coveragerc file to influence files and directories being accounted for in test coverage results

lewisemm committed:
(fix) Remove Travis CI virtualenv from test coverage results.

lewisemm committed:
(chore) Use inheritance in tests to enhance DRY principle

lewisemm committed:
(feature) Create tests for url '/bucketlists/<int:id>/items/<int:item_id>'.

lewisemm committed:
(feature) Add docstring to methods

lewisemm committed:
(fix) Remove unnecessary logic

lewisemm committed:
(feature) Add tests for pagination and 'search by bucketlist name' query string

andela-anandaa commented on api.py:
```diff
@@ -0,0 +1,438 @@
+			return user
+		else:
+			return False
+	else:
```
**Code Quality - Nitpick**
If the `else` block has only on statement (`return` expression), you can do without the `else` key word, thus:

``` python
if user:
    return user
return False
```

andela-anandaa commented:
### README / Comments
- Well written README
- Good: placed Coveralls, TravisCI and QuantifiedCode badges on the README
- Code is well documented / commented, with proper use of inline comments and docStrings

### Functionality
- Nitpick: _Logout_ only tied to timed token, should add functionality of invalidating the token.
- Code works as expected

### Code Quality
- Good code quality
- Good: incorporated [QuantifiedCode](https://www.quantifiedcode.com/app/project/060712fe227e4c2387ebd9fee6b7da22)

### Testing
- 92% code coverage
- Good thing integrating with TravisCI and Coveralls

### File Structure
- The file structure is okay

### Commit Messages
- Consistent and well structured commit messages

lewisemm committed:
(feature) Add 'is_active' field to User model

- Added 'is_active' field to user model to aid in disabling access to resources when the user wants to logout before the token expires.
- Also added docstrings to methods and classes
- Implemented '/auth/login' endpoint of the API

lewisemm committed:
(chore) Eliminate else blocks which process a single line return statement

lewisemm committed:
(feature) Add docstrings to classes

lewisemm committed:
(chore) Update requirements.txt

lewisemm committed:
(chore) Update requirements file

lewisemm committed:
(feature) Add tests for '/auth/logout' endpoint

lewisemm committed:
(chore) Update readme file
